### PR TITLE
Core/adding standard check near

### PR DESCRIPTION
**Pull request by GuillermoCasas**
This pull request adresses #4305. I have used the definition of the tolerance that I proposed there, although I am open to change it to that proposed by @msandre in the same issue.

The advantages of `KRATOS_STANDARD_CHECK_NEAR` are:

- We avoid the proliferation of too many arbitrary tolerances
- We scale the numbers being compared for the user
- We make the instances where the standard check does not work much more conspicuous, implying that the loss of accuracy perhaps cannot be ignored there.

The latter point is obviously dependent upon a generalized used of `KRATOS_STANDARD_CHECK_NEAR.`

I have included a variable Globals::Epsilon ~ 1e-12. The idea is to have it accessible from everywhere and to facilitate a centralized control of this variable. I have also included a first example of use of the new check in test_math_utils.cpp, where all the tolerances where previously set to 1e-6. I have not checked, but I believe it is likely that all checks have become substantially tighter (unless tiny quantities where being compared in some example)

### CODE REVIEW

**Review by roigcarlo — COMMENTED** [review]
Seems good to me, but let me comment it in the tech comm first as lately we are a little bit picky with macro names.

File: kratos/includes/global_variables.h [hunk 1]
```diff
@@ -34,6 +34,7 @@ namespace Kratos
 namespace Globals
 {
 	constexpr double Pi   = 3.14159265358979323846264338327950288419716939937510582097494459230781640628620899862803482534211706798214808651L;
+	constexpr double Epsilon = 1e4 * std::numeric_limits<double>::epsilon();
```
Comment (loumalouomega) [inline kratos/includes/global_variables.h:37]: I remember when I added this some time ago...

**Comment by loumalouomega** [comment]
Ok from my side

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
I don't think this way of checking is correct.
Imagine that you have two numbers:

a = 1e-30
b = 100e-30

there is no problem working with such numbers since they have the same exponent and a and b are very different, nevertheless your check would say that they coincide ...

**Comment by GuillermoCasas** [comment]
> I don't think this way of checking is correct.
> Imagine that you have two numbers:
> 
> a = 1e-30
> b = 100e-30
> 
> there is no problem working with such numbers since they have the same exponent and a and b are very different, nevertheless your check would say that they coincide ...

They are only very different in a relative sense. Any nonzero number will be infinitely different from 0 in a relative sense, so there would be no way of checking if a number is near 0.

The formula I propose assumes that Globals::Epsilon is a small number compared to the precision of the problem. Setting this scale is necessary and cannot be infered from the two numbers provided (they could just be numerical noise). So it is a convention and one should be aware of that when using it. In other words: **this check is valid for problems where the working precision is not too many orders of magnitude smaller than 1**.

If you know you are working at extremely small scales (for very big scales there is no problem), you can either normalise your quantities outside or use `KRATOS_CHECK_NEAR` instead.

In any case, I am of course open to other designs, but I still think it will be useful and clean.

**Comment by RiccardoRossi** [comment]
@GuillermoCasas  the point is exactly what you are telling...they are not zero and the relative difference is large. I don't think we can have a "default" function for doing this, in the sense that we should have the possibility of giving some user control to this.

now, your approach as surely some merits, in that it allows to compare numbers in a more consistent way.
As i understand your formula is (assuming a and b are positive, otherwise with absolute values)

     (a - b)/(1 + a + b) < 10000*numerical_epsilon

aside for the 10000 that looks a little arbitrary to me, one option could be to modify this as

     (a-b)/(reference + a + b) < factor*numerical_epsilon

where reference is a user-specified value (for example in my example you would set it to 1e-30)

regarding the "factor" i also think it should be user-specifyable (although most probably 10k could be employed as default)

**Comment by GuillermoCasas** [comment]
> @GuillermoCasas the point is exactly what you are telling...they are not zero and the relative difference is large. I don't think we can have a "default" function for doing this, in the sense that we should have the possibility of giving some user control to this.
> 
> now, your approach as surely some merits, in that it allows to compare numbers in a more consistent way.
> As i understand your formula is (assuming a and b are positive, otherwise with absolute values)
> 
> ```
>  (a - b)/(1 + a + b) < 10000*numerical_epsilon
> ```
> aside for the 10000 that looks a little arbitrary to me, one option could be to modify this as
> 
> ```
>  (a-b)/(reference + a + b) < factor*numerical_epsilon
> ```
> where reference is a user-specified value (for example in my example you would set it to 1e-30)
> 
> regarding the "factor" i also think it should be user-specifyable (although most probably 10k could be employed as default)

Sure, 10K is completely arbitrary. My criterion was to be comfortably far from the std epsilon so as to avoid spurious false negatives, but it could have been 100, 1000, or some other number.

I am not advocating for eliminating the possibility to set your constant. I am just suggesting that this is unnecessary in most cases and it is only in exceptional occasions that we will need to do so. The proof is that if you check the tolerances being used around in Kratos, those are just numbers that 'look small'. In fact, this is often not really problematic because people tend to work around order-1 quantities (+- a few orders of magnitude).

Moreover, the relative difference is not necessarily relevant! It is only relevant when the numbers involved happen to be of the characteristic scale of the problem (L). You are right that the correct epsilon should be proportional to L. What I am doing is partially correcting for this scale when the numbers are large: perhaps they are not large with respect to L but at least I take their size into account.

In rare cases, it could happen that all the numbers involved are much larger than L, which would amount to having all your measures shifted by several orders of magnitude. In that case, scaling by the size of the numbers as I suggested would be wrong.

However, I am starting to think that perhaps is would be better to avoid trying to infer any information about L from the numbers provided and simply comparing the absolute difference to Globals::Epsilon would be less problematic.

In summary, I agree that the solution is far from perfect, and I will study it further, but I want to insist in the following points:

- Forcing developers to add a tolerance will not in practice help having tighter and more consistent checks
- The relative error is not relevant, since the target needs not be relatively close to L (in fact, it is often 0)

File: kratos/includes/checks.h [hunk 2]
```diff
@@ -42,6 +42,9 @@
 #define KRATOS_CHECK_STRING_CONTAIN_SUB_STRING(TheString, SubString) if (TheString.find(SubString) == std::string::npos ) \
 KRATOS_ERROR << "The string \"" << SubString << "\" was not found in the given string"
 
+#define KRATOS_STANDARD_CHECK_NEAR(a,b) if(std::abs(a - b) > Globals::Epsilon * (1 + std::abs(a) + std::abs(b))) KRATOS_ERROR << "Check failed because " << #a << " = " << a << \
```
Comment (msandre) [inline kratos/includes/checks.h:57]: The check is useful for more than just tests. So it would make sense to split the check from the exception. You could put the check into a regular function, e.g., `bool IsNear(double a, double b);` in the Kratos namespace and call it here.

File: kratos/includes/global_variables.h (hunk 1 above)
Comment (msandre) [reply kratos/includes/global_variables.h:37]: I'm not sure how useful this is as a global.

**Review by msandre — COMMENTED** [review]
It may be useful to define a separate function to check if a number is close to zero. I would also agree with keeping the function parameters simple.

File: kratos/includes/checks.h (hunk 2 above)
Comment (GuillermoCasas) [reply kratos/includes/checks.h:57]: Sure, I will do that. Thanks

Comment (GuillermoCasas) [reply kratos/includes/checks.h:57]: About the definition of Epsilon as a global, I just found it appealing to have the arbitrary constant centralised (though it needs not be fixed to the value I proposed). Rationalising my decision though, I can see the following advantages:

First, it makes it clear that a convention has been reached about a standard small value in Kratos, (at least partially) discouraging further ad-hoc definitions of the kind. 

Second, it becomes an accessible variable for everyone to use. For instance, it could be reused to create new macros etc.

Admittedly, these are not super strong reasons so let's see what people think.

**Comment by loumalouomega** [comment]
From my side I support this, I did something similar some time ago.

File: kratos/includes/global_variables.h (hunk 1 above)
Comment (pooyan-dadvand) [reply kratos/includes/global_variables.h:37]: Where this expression come form?

**Comment by loumalouomega** [comment]
@GuillermoCasas can we delete the branch?
